src/typeSystem.js

```diff
@@ -52,52 +40,58 @@ class Type {
   }
 }
 
-class FunctionType extends Type {
-  constructor(inputType, outputType) {
+class EstructuralType extends Type {
+  constructor(name, attributes) {
     super();
-    this.inputType = inputType;
-    this.outputType = outputType;
+    this.name = name
+    this.attributes = attributes
   }
 
-  isFunctionType() {
-    return true
-  }
+  attributeValues() { return Object.values(this.attributes) }
 
   allTypeVariableNamesUsed() {
-    return this.inputType.allTypeVariableNamesUsed().concat(this.outputType.allTypeVariableNamesUsed())
+    return this.attributeValues().flatMap(type => type.allTypeVariableNamesUsed())
   }
 
-  isSameType(otherType) {
-    return otherType.inputType &&
-      otherType.outputType &&
-      this.inputType.isSameType(otherType.inputType) &&
-      this.outputType.isSameType(otherType.outputType)
+  includes(parametricType) { //TODO: Revisar esto
+    const typeVariableName = parametricType.typeVariableName
+    return this.attributeValues().some(type => type.typeVariableName == typeVariableName || type.includes(parametricType))
   }
 
-  includes(parametricType) {
-    const typeVariableName = parametricType.typeVariableName
-    return (this.inputType.typeVariableName == typeVariableName || this.outputType.typeVariableName == typeVariableName) ||
-      this.inputType.includes(parametricType) || this.outputType.includes(parametricType)
+  restrictToSimple(type) {
+    throw typeError(type, this)
   }
 
-  replacing(typeMap) {
-    return new FunctionType(this.inputType.replacing(typeMap), this.outputType.replacing(typeMap))
+  restrictToComposite(type) { //TODO: Mover a constraints
+    const constraints = Object.values(zipObjects(type.attributes, this.attributes)).flatMap(([type1, type2]) => type1.eqConstraints(type2))
```

Comment:
Not sure I follow here
What happens if zipObjects between different types of composite types ((a -> b) vs. (a, b)) behave? It returns an empty list and the error is caught on `constraints.length` below? :thinking: 

Also, with the code as it is now, is there a rule of how many constraints can come from a function type? (Like, as many constraints as the number of input types + output type). Because otherwise I think the length == attributeValues().length could cause problems

Reply:
Sí, esto quedó medió ahí. Para el primer caso que mencionás tenía pensado matchear por `name` también (pensé que estaba hecho). Lo segundo que decís no terminé de entender por qué lo decís.

Ahora pensándolo mejor propongo validar que que los tipos sean compatibles que es cuando:
- Tienen el mismo name
- Tienen las mismas `attributes` keys

¿Qué te pa?

Reply:
Ahí lo cambié.

Reply:
me agrada, ahi veo el codigo :eyes: